pkg/reportingplugins/mercury/v1/mercury.go

```diff
@@ -0,0 +1,428 @@
+	var pao parsedAttributedObservation
+	var obs MercuryObservationProto
+	if err := proto.Unmarshal(ao.Observation, &obs); err != nil {
+		return parsedAttributedObservation{}, pkgerrors.Errorf("attributed observation cannot be unmarshaled: %s", err)
+	}
+
+	pao.Timestamp = obs.Timestamp
+	pao.Observer = ao.Observer
+
+	if obs.PricesValid {
+		var err error
+		pao.BenchmarkPrice, err = mercury.DecodeValueInt192(obs.BenchmarkPrice)
+		if err != nil {
+			return parsedAttributedObservation{}, pkgerrors.Errorf("benchmarkPrice cannot be converted to big.Int: %s", err)
+		}
+		pao.Bid, err = mercury.DecodeValueInt192(obs.Bid)
+		if err != nil {
+			return parsedAttributedObservation{}, pkgerrors.Errorf("bid cannot be converted to big.Int: %s", err)
+		}
+		pao.Ask, err = mercury.DecodeValueInt192(obs.Ask)
+		if err != nil {
+			return parsedAttributedObservation{}, pkgerrors.Errorf("ask cannot be converted to big.Int: %s", err)
+		}
+		pao.PricesValid = true
+	}
+
+	if obs.MaxFinalizedTimestampValid {
+		pao.MaxFinalizedTimestamp = obs.MaxFinalizedTimestamp
+		pao.MaxFinalizedTimestampValid = true
+	}
+
+	if obs.LinkFeeValid {
+		var err error
+		pao.LinkFee, err = mercury.DecodeValueInt192(obs.LinkFee)
+		if err != nil {
+			return parsedAttributedObservation{}, pkgerrors.Errorf("link price cannot be converted to big.Int: %s", err)
+		}
+		pao.LinkFeeValid = true
+	}
+	if obs.NativeFeeValid {
+		var err error
+		pao.NativeFee, err = mercury.DecodeValueInt192(obs.NativeFee)
+		if err != nil {
+			return parsedAttributedObservation{}, pkgerrors.Errorf("native price cannot be converted to big.Int: %s", err)
+		}
+		pao.NativeFeeValid = true
+	}
+
+	return pao, nil
+}
+
+func parseAttributedObservations(lggr logger.Logger, aos []ocrtypes.AttributedObservation) []ParsedAttributedObservation {
+	paos := make([]ParsedAttributedObservation, 0, len(aos))
+	for i, ao := range aos {
+		pao, err := parseAttributedObservation(ao)
+		if err != nil {
+			lggr.Warnw("parseAttributedObservations: dropping invalid observation",
+				"observer", ao.Observer,
+				"error", err,
+				"i", i,
+			)
+			continue
+		}
+		paos = append(paos, pao)
+	}
+	return paos
+}
+
+func (rp *reportingPlugin) Report(repts ocrtypes.ReportTimestamp, previousReport ocrtypes.Report, aos []ocrtypes.AttributedObservation) (shouldReport bool, report ocrtypes.Report, err error) {
+	paos := parseAttributedObservations(rp.logger, aos)
+
+	// By assumption, we have at most f malicious oracles, so there should be at least f+1 valid paos
+	if !(rp.f+1 <= len(paos)) {
+		return false, nil, pkgerrors.Errorf("only received %v valid attributed observations, but need at least f+1 (%v)", len(paos), rp.f+1)
+	}
+
+	observationTimestamp := mercury.GetConsensusTimestamp(Convert(paos))
+
+	var validFromTimestamp uint32
+	if previousReport != nil {
+		validFromTimestamp, err = rp.reportCodec.ObservationTimestampFromReport(previousReport)
```

Comment:
Does it not need 1 second added?

Reply:
I don't think so, why? We make sure that `validFromTimestamp` is not greater than current `observationTimestamp`